File: 1601_maximum_number_of_achievable_transfer_requests/solution.go
// LeetCode 1601 - Maximum Number of Achievable Transfer Requests
// https://leetcode.com/problems/maximum-number-of-achievable-transfer-requests/

func maximumRequests(n int, requests [][]int) int {
	ans := 0
	m := len(requests)
	for mask := 0; mask < 1<<m; mask++ {
		cnt := 0
		for t := mask; t > 0; t &= t - 1 {
			cnt++
		}
		if cnt <= ans {
			continue
		}
		bal := make([]int, n)
		for i, req := range requests {
			if mask>>i&1 == 1 {
				bal[req[0]]--
				bal[req[1]]++
			}
		}
		ok := true
		for _, v := range bal {
			if v != 0 {
				ok = false
				break
			}
		}
		if ok {
			ans = cnt
		}
	}
	return ans
}

File: 1601_maximum_number_of_achievable_transfer_requests/solution.rs
// LeetCode 1601 - Maximum Number of Achievable Transfer Requests
// https://leetcode.com/problems/maximum-number-of-achievable-transfer-requests/

impl Solution {
    pub fn maximum_requests(n: i32, requests: Vec<Vec<i32>>) -> i32 {
        let m = requests.len();
        let mut ans = 0;
        for mask in 0u32..(1u32 << m) {
            let cnt = mask.count_ones() as i32;
            if cnt <= ans {
                continue;
            }
            let mut bal = vec![0i32; n as usize];
            for (i, req) in requests.iter().enumerate() {
                if mask >> i & 1 == 1 {
                    bal[req[0] as usize] -= 1;
                    bal[req[1] as usize] += 1;
                }
            }
            if bal.iter().all(|&v| v == 0) {
                ans = cnt;
            }
        }
        ans
    }
}
